### PR TITLE
[Downgrade 7.0] Fix value calls that do not lex or lex differently in PHP 5.6

### PR DESCRIPTION
**Pull request by jtojnar**
PHP 7.0 implemented [Uniform Variable Syntax RFC](https://wiki.php.net/rfc/uniform_variable_syntax), which allows many expressions that did not previously parse and changes parsing of some less common ones.

### CODE REVIEW

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                ,
+                <<<'CODE_SAMPLE'
+final class Foo
+{
+    /** @var callable handler for sending emails */
+    public $mailer;
+}
+
+$foo = new Foo;
+($foo->mailer)($args);
```
Comment (TomasVotruba) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]:
```suggestion
```

Possibly left over

**Comment by TomasVotruba** [comment]
Thank you, very clean work :+1:

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php (hunk 1 above)
Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]: Right, thanks. I fixed it in the tests but forgot about the example.

**Comment by jtojnar** [comment]
Hmm, turns out tracy also uses IIFE:

```
(function() {
    echo 'foo';
})();
```

and

`($this->resourceExposers[$type])($resource)`

which really makes no sense to me but we should probably extend it to all expression that are not expressible without parentheses. But I am not sure how to find those out – maybe it is okay to add them on case by case basis.

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public function refactor(Node $node): ?FuncCall
+    {
+        if (! $node->name instanceof PropertyFetch) {
+            return null;
+        }
+
+        $args = array_merge([new Arg($node->name)], $node->args);
+
+        return new FuncCall(new Name('call_user_func'), $args);
```
Comment (samsonasik) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]: You can use NodeFactory service

https://github.com/rectorphp/rector-src/blob/b76243b41682edd072da3ca149461283c779415b/src/PhpParser/Node/NodeFactory.php#L417

The property $nodeFactory is available from `AbstractRector` already:

https://github.com/rectorphp/rector-src/blob/b76243b41682edd072da3ca149461283c779415b/src/Rector/AbstractRector.php#L88

Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]: I guess we need a rector for that :rofl: since I copied it from another rector :rofl:

Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]: Not sure if we need it though, since the arguments are already `Arg[]`. Also what is the difference between `NodeFactory::createFuncCall` and `BuilderFactory::funcCall`?

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        );
+    }
+
+    private function isLexedDifferentlyInPhp5(Expr $expr): bool
+    {
+        return (
+            // PHP 5.6 parses `self::$foo['test']()` as a static method call
+            // `self::{$foo['test']}()` rather than `(self::$foo)['test']()` like PHP 7.0 does.
+            // See php5-static-property-weirdness.php.inc fixture.
+            $expr instanceof ArrayDimFetch && $expr->var instanceof StaticPropertyFetch
```
Comment (jtojnar) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php]: Same parsing difference occurs with `self::$foo["test"]["foo"]();`, which could be captured by `$expr instanceof ArrayDimFetch && $expr->var instanceof ArrayDimFetch && $expr->var->var instanceof StaticPropertyFetch` and ad infinitum.

I used the following script to compare the parsing differences:

```php
<?php

use PhpParser\ParserFactory;
use PhpParser\NodeDumper;

require_once __DIR__ . '/vendor/autoload.php';

$data = '<?php self::$foo["test"]["foo"]();';

$parserFactory = new ParserFactory();
$nodeDumper = new NodeDumper();
// $phpParser = $parserFactory->create(ParserFactory::ONLY_PHP5);
$phpParser = $parserFactory->create(ParserFactory::PREFER_PHP7);
$stmts = $phpParser->parse($data);
echo $nodeDumper->dump($stmts[0]->expr) . PHP_EOL;
```

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            // The first function call does not even need to be wrapped in parentheses
+            // but PHP 5 still does not like curried functions like `f($args)($moreArgs)`.
+            || $expr instanceof FuncCall
+        );
+    }
+
+    private function isLexedDifferentlyInPhp5(Expr $expr): bool
+    {
+        return (
+            $this->isParsedAsStaticCallNameByPhp5($expr)
```
Comment (jtojnar) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php]: The semantic changes should probably go to a different rector.

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php (hunk 2 above)
Comment (samsonasik) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeParenthesizedValueCallToCallUserFuncRector.php]: both are helper shortcut to create FuncCall, `BuilderFactory::funcCall` probably too heavy for a certainly passed string name and arg.

File: rules/DowngradePhp70/Rector/Expression/DowngradeUnnecessarilyParenthesizedExpressionRector.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\DowngradePhp70\Expression;
```
Comment (jtojnar) [inline rules/DowngradePhp70/Rector/Expression/DowngradeUnnecessarilyParenthesizedExpressionRector.php:5]: PHPStan is little overzealous:

  - '#Class "Rector\\DowngradePhp70\\Rector\\Expression\\DowngradeUnnecessarilyParenthesizedExpressionRector" has invalid namespace category "Expression"\. Pick one of\: "ArrayDimFetch", PropertyFetch", MethodCall", StaticPropertyFetch", StaticCall", FuncCall"#'

It forces me to pick a single one when I match all of them.

File: rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        // TODO: Make PHPStan allow Expr namespace.
+        $dummyClassNameUse = str_contains(Expr::class, '\\');
+        assert($dummyClassNameUse);
```
Comment (jtojnar) [inline rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php:61]: Ugh we need to add `Expr` into https://github.com/rectorphp/phpstan-rules/blob/df714324fd2839c9f03d39c437fa6b37ccf127a3/src/Rule/RequireRectorCategoryByGetNodeTypesRule.php#L124. Also discovered https://github.com/symplify/symplify/issues/3872 while working around this.

File: rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        // since PHP 5 cannot deal with them.
+        // In some cases parens are necessary so they will remain even after this rule is applied
+        // but other rules should handle that.
+        return match (true) {
+            $node instanceof ArrayDimFetch => new ArrayDimFetch($node->var, $node->dim),
+            $node instanceof PropertyFetch => new PropertyFetch($node->var, $node->name),
+            $node instanceof MethodCall => new MethodCall($node->var, $node->name, $node->args),
+            $node instanceof StaticPropertyFetch => new StaticPropertyFetch($node->class, $node->name),
+            $node instanceof StaticCall => new StaticCall($node->class, $node->name, $node->args),
+            default /* $node instanceof FuncCall */ => new FuncCall($node->name, $node->args),
```
Comment (jtojnar) [inline rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php]: Another PHPStan meh.

Comment (TomasVotruba) [reply rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php]: PHPStan is right here, as this technically works, but `match (true)` magic is not suitable for type matching. 

Here would be better collector pattern as we have e.g. in `NodeNameResolver`.

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://wiki.php.net/rfc/uniform_variable_syntax
+ *
+ * @see \Rector\Tests\DowngradePhp70\Rector\FuncCall\DowngradeUncallableValueCallToCallUserFuncRector\DowngradeUncallableValueCallToCallUserFuncRectorTest
+ */
+final class DowngradeUncallableValueCallToCallUserFuncRector extends AbstractRector
```
Comment (TomasVotruba) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php:26]: Could you split this rule to own PR? It seems it's ready and it would lower complexity for further review.

Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php:26]: Moved the unfinished rule to #1617.

File: rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php (hunk 7 above)
Comment (jtojnar) [reply rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php]: Could you be more specific? I do not see anything applicable in `NodeNameResolver` and collector pattern does not return any relevant results.

Comment (TomasVotruba) [reply rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php]: The `NodeNameResolver` accepts array of `array $nodeNameResolvers = []` that each resolve own types based on instanceof.
Before we had this huge single class that contained everything. This now allows to focus on one node at a time.

**Review by TomasVotruba — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Looks good to me :+1: 

@samsonasik Feel free to merge when you thinks its ready.

File: rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        return (
+            // Interpreted as MethodCall without parentheses.
+            $expr instanceof PropertyFetch
+            // Interpreted as StaticCall without parentheses.
+            || $expr instanceof StaticPropertyFetch
+            || $expr instanceof Closure
+            // The first function call does not even need to be wrapped in parentheses
+            // but PHP 5 still does not like curried functions like `f($args)($moreArgs)`.
+            || $expr instanceof FuncCall
```
Comment (samsonasik) [inline rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php]: I prefer to use `in_array()` with defined nodes in constant, eg:

```php
/** @var @var array<class-string<Expr>> */
private const INDIRECT_CALLABLE_EXPR = [
   // Interpreted as MethodCall without parentheses.
    PropertyFetch::class,
   // Interpreted as StaticCall without parentheses.  
    StaticPropertyFetch::class,
    Closure::class,
    // The first function call does not even need to be wrapped in parentheses
    // but PHP 5 still does not like curried functions like `f($args)($moreArgs)`.
    FuncCall::class,
];

// ...
return in_array($expr::class, self::INDIRECT_CALLABLE_EXPR, true);
```

Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php]: Hmm, that should work if none of the expressions have subclasses.

Comment (jtojnar) [reply rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php]: Did that.

**Comment by samsonasik** [comment]
Thank you @jtojnar
